lib/Source/AutoHog/AutoHogImporter.cpp

```diff
@@ -0,0 +1,516 @@
+    } else if (strcmp(cellType, "NAND") == 0) {
+      op = builder.create<cggi::NandOp>(funcOp.getLoc(), resultTypes, operands);
+    } else if (strcmp(cellType, "NOR") == 0) {
+      op = builder.create<cggi::NorOp>(funcOp.getLoc(), resultTypes, operands);
+    } else if (strcmp(cellType, "OR") == 0) {
+      op = builder.create<cggi::OrOp>(funcOp.getLoc(), resultTypes, operands);
+    } else if (strcmp(cellType, "XOR") == 0) {
+      op = builder.create<cggi::XorOp>(funcOp.getLoc(), resultTypes, operands);
+    } else if (strcmp(cellType, "XNOR") == 0) {
+      op = builder.create<cggi::XNorOp>(funcOp.getLoc(), resultTypes, operands);
+    } else if (strcmp(cellType, "NOT") == 0) {
+      op = builder.create<cggi::NotOp>(funcOp.getLoc(), resultTypes, operands);
+    } else {
+      assert(false && "Unknown cell type");
+    }
+
+    cellsByName[cellName] = op;
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << "Ops created for circuit cells. Func (which "
+                             "should not verify because it has no return): "
+                          << funcOp << "\n");
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "cellOutputPortToResultIndex=\n";
+    for (auto &[cellName, portToIndex] : cellOutputPortToResultIndex) {
+      llvm::dbgs() << "  " << cellName << ":\n";
+      for (auto &[portName, index] : portToIndex) {
+        llvm::dbgs() << "    " << portName << " -> " << index << "\n";
+      }
+    }
+  });
+
+  SmallVector<Value> outputValues;
+  outputValues.resize(numOutputs);
+  for (auto &[portName, port] : ioPorts) {
+    if (port.type != PortType::OUTPUT) continue;
+    LLVM_DEBUG(llvm::dbgs()
+               << "Finding value for output port: " << portName << "\n");
+
+    if (port.connectedInputPort.has_value()) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "Output port " << portName << " is connected to input port "
+                 << port.connectedInputPort.value() << "\n");
+      auto indexOp = builder.create<arith::ConstantIndexOp>(
+          funcOp.getLoc(),
+          inputPortToTensorIndex[port.connectedInputPort.value()]);
+      outputValues[outputPortToTensorIndex[portName]] =
+          builder.create<tensor::ExtractOp>(funcOp.getLoc(), funcArg,
+                                            indexOp.getResult());
+      continue;
+    }
+
+    if (outputPortToConnectedCellPort.contains(portName)) {
+      Port cellPort = outputPortToConnectedCellPort[portName];
+      LLVM_DEBUG(llvm::dbgs()
+                 << "Output port " << portName << " is connected to cell port "
+                 << cellPort.portName << " of cell "
+                 << cellPort.cellName.value_or("") << "\n");
+      Operation *op = cellsByName[cellPort.cellName.value().c_str()];
+      int resultIndex =
+          cellOutputPortToResultIndex.at(cellPort.cellName.value_or(""))
+              .at(cellPort.portName);
+      outputValues[outputPortToTensorIndex[portName]] =
+          op->getResult(resultIndex);
+      continue;
+    }
+  }
```

Comment:
should there also be a JSON failure here detecting if there was no connection to that output port? i guess, to me there's also a possibility that an output port is a constant, which i don't think would appear in either of these two cases.

Reply:
Done